clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3070,4 +3079,235 @@ void handleFailure(Throwable throwable) {
+                    for (Map.Entry<Integer, Optional<NewPartitionReassignment>> partitionEntry :
+                            partitionsToReassignments.entrySet()) {
+                        int partitionIndex = partitionEntry.getKey();
+                        Optional<NewPartitionReassignment> reassignment = partitionEntry.getValue();
+
+                        ReassignablePartition reassignablePartition = new ReassignablePartition()
+                                .setPartitionIndex(partitionIndex)
+                                .setReplicas(reassignment.map(NewPartitionReassignment::targetBrokers).orElse(null));
+                        reassignablePartitions.add(reassignablePartition);
+                    }
+
+                    ReassignableTopic reassignableTopic = new ReassignableTopic()
+                            .setName(topicName)
+                            .setPartitions(reassignablePartitions);
+                    data.topics().add(reassignableTopic);
+                }
+                data.setTimeoutMs(timeoutMs);
+                return new AlterPartitionReassignmentsRequest.Builder(data);
+            }
+
+            @Override
+            public void handleResponse(AbstractResponse abstractResponse) {
+                AlterPartitionReassignmentsResponse response = (AlterPartitionReassignmentsResponse) abstractResponse;
+                Map<TopicPartition, ApiException> errors = new HashMap<>();
+                int expectedResponsesCount = topicsToReassignments.values().stream().mapToInt(Map::size).sum();
+                int receivedResponsesCount = 0;
+
+                Errors topLevelError = Errors.forCode(response.data().errorCode());
+                switch (topLevelError) {
+                    case NONE:
+                        break;
+                    case NOT_CONTROLLER:
+                        handleNotControllerError(topLevelError);
+                        break;
+                    default:
+                        for (ReassignableTopicResponse topicResponse : response.data().responses()) {
+                            String topicName = topicResponse.name();
+                            for (ReassignablePartitionResponse partition : topicResponse.partitions()) {
+                                errors.put(
+                                        new TopicPartition(topicName, partition.partitionIndex()),
+                                        new ApiError(topLevelError, topLevelError.message()).exception()
+                                );
+                                receivedResponsesCount += 1;
+                            }
+                        }
+                        break;
+                }
+                if (topLevelError == Errors.NONE)
+                    receivedResponsesCount += validateTopicResponses(response.data().responses(), errors);
+
+                if (errors.values().stream().noneMatch(Objects::nonNull)
```

Comment:
I can't come up with a way that makes this better to read. Could you give a concrete example?

Reply:
It would be something like this:
```
private void assertResponseCountMatch(Map<TopicPartition, ApiException> errors, int expectedResponsesCount, int receivedResponsesCount) {
                if (errors.values().stream().noneMatch(Objects::nonNull)
                        && receivedResponsesCount != expectedResponsesCount) {
                    String quantifier = receivedResponsesCount > expectedResponsesCount ? "many" : "less";
                    throw new UnknownServerException("The server returned too " + quantifier + " results." +
                            "Expected " + expectedResponsesCount + " but received " + receivedResponsesCount);
                }
            }
```

Reply:
Ah, just that block? Sure, let's do it.

I generally like encapsulating small chunks like this into methods because it makes the code waay easier to read.
Yet, I believe I've had it questioned in some reviews in terms of "why have this in a method if it's used only once"

Glad to have the opposite recommended :)

Reply:
Yea I can understand the argument as I often debate it during development but the JIT compiler would inline these kind of methods so you actually don't lose any performance yet the code looks more understandable :)

Reply:
I don't think performance was the reason I've got such remarks but yeah, I agree it looks better